src/_guides/language/effective-dart/design.md

```diff
@@ -974,79 +1060,260 @@ var names = people.map((Person person) {
 });
 {% endprettify %}
 
+In rare cases, the surrounding context is not precise enough to provide a type
+for one or more of the function's parameters. In those cases, you may need to
+annotate.
+
 
-### AVOID annotating with `dynamic` when not required.
+### PREFER annotating with `dynamic` even when not required.
 
-{% comment %}
-update-for-dart-2.0
-https://github.com/dart-lang/site-www/pull/480#discussion_r157355593
-{% endcomment %}
+When you omit a type annotation, *most* of the time, Dart infers a correct
+static type for you. That word "most" is the alarming part. Sometimes, inference
+fails and silently uses `dynamic` instead. If you didn't realize that inference
+failed, you now have code that *looks* statically safe but actually has no
+type checking whatsoever.
 
-In most places in Dart, a type annotation can be omitted, in which case the type
-will automatically be `dynamic`. Thus, omitting the type annotation entirely is
-semantically equivalent but more terse.
+Sometimes, `dynamic` is what you want. Dart does support a dynamic type for good
+reasons. But a casual reader of your code who sees an annotation is missing has
```

Comment:
> ... for good reason.

@kwalrath WDYT?

Reply:
"for good reasons" sounds more natural to me.

("for good reason" is the actual saying, I think, but it sounds formal.)